### PR TITLE
Refactor PlayerEventHandler

### PR DESCRIPTION
**Pull request by frost-byte**
Added a helper class for categorizing blocks not covered by the Tag system
Heavily Refactored PlayerInteractEvent handling into multiple functions
Some formatting and whitespace cleanup

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,7 +48,7 @@ Replace `<version>` number with this number: [![](https://jitpack.io/v/TechFortr
 	<dependency>
 	    <groupId>com.github.TechFortress</groupId>
 	    <artifactId>GriefPrevention</artifactId>
-	    <version>16.7.1</version>
+	    <version>16.9</version>
```
Comment (RoboMWM) [inline README.md:51]: Please do not change this.

File: pom.xml [hunk 2]
```diff
@@ -68,6 +68,13 @@
                 <artifactId>bukkit</artifactId>
                 <version>1.13-R0.1-SNAPSHOT</version>
                 <scope>provided</scope>
+        </dependency>
+        <!--Worldedit dependency-->
```
Comment (RoboMWM) [inline pom.xml:72]: Is this needed?

File: src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java [hunk 3]
```diff
@@ -3475,13 +3508,20 @@ public void restoreChunk(Chunk chunk, int miny, boolean aggressiveMode, long del
...
+		World world = chunk.getWorld();
+
 		for(int x = 0; x < snapshots.length; x++)
 		{
 			for(int z = 0; z < snapshots[0][0].length; z++)
 			{
 				for(int y = 0; y < snapshots[0].length; y++)
 				{
-					Block block = chunk.getWorld().getBlockAt(startLocation.getBlockX() + x, startLocation.getBlockY() + y, startLocation.getBlockZ() + z);
+					block = world.getBlockAt(startX + x, startY + y, startZ + z);
```
Comment (RoboMWM) [inline src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java:3524]: Don't really see any benefit here other than maybe a slight boost in readability, especially since this data isn't used again anywhere else.

File: src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java [hunk 4]
```diff
@@ -96,19 +97,27 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.regex.Pattern;
 
-class PlayerEventHandler implements Listener 
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+@SuppressWarnings({
```
Comment (RoboMWM) [inline src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java:103]: why?

**Review by RoboMWM — CHANGES_REQUESTED** [review]
So a bit more than just PlayerEventHandler was refactored, but this really seems more like a style change than anything else. Perhaps if you were revising the logic it'd make sense. 

It also seems like your work is missing the last several commits that've been pulled in considering the diff in some areas?

A lot of refactoring and rewriting is being done in version 20.

File: README.md (hunk 1 above)
Comment (frost-byte) [reply README.md:51]: Alright, I'll revert that.  Didn't recall changing it.

File: pom.xml (hunk 2 above)
Comment (frost-byte) [reply pom.xml:72]: I'll take this out as well

File: src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java (hunk 3 above)
Comment (frost-byte) [reply src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java:3524]: How many elements are there total in the snapshots array? ~82000?
Does it really make sense to call startLocation.getBlockX, startLocation.getBlockZ that many times when the value doesn't change?
You don't need to redeclare the Block object every time that inner loop is called either.
So while this isn't a functional change, it's an optimization.

File: src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java (hunk 4 above)
Comment (frost-byte) [reply src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java:103]: Are you asking how these are beneficial? or do you think it's better to have a bunch of null checks in each function?  They make it clearer to anyone using the function that the parameters cannot be null, or can potentially be null.

**Comment by frost-byte** [comment]
That one function was trying to do so much it was hard to follow. While refactoring it doesn't necessarily offer anything functionally, in terms of maintaining and working with the code, you gain a lot of benefit.
You reduce a lot of repetition and add some clarity.   Obviously you've been managing with what you had, but that made it very difficult to add something.

Unfortunately some of the differences were inconsistencies in white space, trailing tabs, a mix of tabs and spaces.  In some cases how you format is just a matter of whatever your coding standards are for the repository.  If I overlooked something where you stated what that was, then I apologize for that.

File: src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java (hunk 4 above)
Comment (RoboMWM) [reply src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java:103]: sorry, meant all the warning suppressions. As for the annotations, I recall CB/spigot servers having issues with these for some reason...

File: src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java (hunk 3 above)
Comment (RoboMWM) [reply src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java:3524]: Oh ok, I see that it's in a loop and that getBlockX runs a floor function.

**Comment by RoboMWM** [comment]
The inconsistencies with tabs and whitespace are because Big_Scary used tabs, and PRs used spaces. So it's everywhere, not just here; and probably best to be done in a separate PR since I'm sure it touches a lot of areas.

As for actual refactoring - that's why version 20 is a thing (along with cleaning up a lot of extraneous features that aren't easily supported/debugged). Hence why I'd suggest and prefer any refactoring work to go into that, instead of the existing 16.x versions. Especially with the large diff in this PR it takes time to review PRs for 16.x, some of which have caused problems because of things I haven't noticed. In other words, it's best to do one change per PR rather than lumping everything together in a single PR and using a general term (a la "refactoring", etc.)

**Comment by frost-byte** [comment]
I had forked the repo last month when spigot was in prerelease and realized that WorldGuard and WorldEdit were being updated.  It was only within the last week that I had noticed some activity on the repo and wanted to pull in the upstream changes that I noticed some issues with my changes in the PlayerInteractEvent listener.  (Which was mostly meant to make material categories). I noticed the bukkit Tag system seems to solve most of that.

Between the whitespace and the massive and difficult to follow function, the refactoring was the only way I could make sense of it.

Nevertheless I do understand where you’re coming from, and I thought I had merged in the most recent commits.  I did test every aspect of what I could see in that function to make sure nothing had been broken.  

I had issues with Restoring Nature in my version as well as the current version.  RestoreFill seems to function, Im just not certain how the other restore methods are supposed to work, other than looking through the blocks in the chunk.  Is this a matter of configuration? Enabling snapshots? Is the chunk supposed to regenerate?

Sorry for the long response, hope you’re able to use some aspects of this pr, and I’m sure you’ll notice a big quality of life improvement once you’re able to refactor everything for v20.

**Comment by RoboMWM** [comment]
> Is the chunk supposed to regenerate?

No, I know that for certain. Big_Scary partly gives the reason for that in the plugin overview on spigot and bukkitdev being that it is "dangerous" and doesn't fix vanilla terrain generation issues. It also will fully avoid any claim. Regardless, due to its constant need to be maintained, I'm taking this out in v20.

> hope you’re able to use some aspects of this pr

That's probably what I'll do. Whitespace issues, which I assume is the majority of this PR diff-wise, is throughout the project, not just in the 13 files you changed - though if you do want to fix that as well, go ahead and make a new PR with just that on the latest (or perhaps at least after 16.9 is released since there are a few pending PRs).

**Comment by frost-byte** [comment]
Most of the things that were not related to white space, were the refactoring of the PlayerInteract Event - dividing it out into different functions.  So I tried to keep my edits confined to that code and the surrounding code. I understand it makes it more difficult to compare and merge in any recent additions that are being made in other pr or updates you make.

**Comment by RoboMWM** [comment]
Yea, best if you can PR to one class at a time for doing stuff like this, else you'll easily get conflicts with other recently-pushed commits. (This PR inevitably has conflicts now.)
